### PR TITLE
Support for ValueResult as public type to carry value and location

### PR DESCRIPTION
**Pull request by KathleenDollard**
I am having an issue where the base of this merge is not what I am expecting. You can see the problem in System.CommandLine.Subsystems/Directives/DirectiveSubsystem.cs which exists in dotnet/command-line-api, but appears as completely fresh code in this PR. Another example is that ParserTests.cs has a bunch of whitespace fixes made by @mhutch that do not appear in the base of this PR. 

I either did something dumb or I have lost my mind.

Commits in this PR also need to be cleaned up, so it is not ready to go, but I wanted to put this out so I could get help resolving this.

### CODE REVIEW

File: src/System.CommandLine.Subsystems.Tests/AlternateSubsystems.cs [hunk 1]
```diff
@@ -1,7 +1,9 @@
 ﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.CommandLine.Directives;
```
Comment (KathleenDollard) [inline src/System.CommandLine.Subsystems.Tests/AlternateSubsystems.cs:4]: This is an example of my confusion. If I look at dotnet:main-powderhouse, I see this line present, as well as the files DiagramSubsystemTests.cs, etc.

File: src/System.CommandLine.Tests/ParserTests.cs [hunk 2]
```diff
@@ -508,119 +510,120 @@ public void Original_order_of_tokens_is_preserved_in_ParseResult_Tokens(string c
...
         {
-            var command = new CliCommand("outer")
-            {
-                new CliArgument<string[]>("arg1"),
-                new CliCommand("inner")
-                {
-                    new CliArgument<string[]>("arg2")
-                }
-            };
+            var command = new CliCommand(
```
Comment (KathleenDollard) [inline src/System.CommandLine.Tests/ParserTests.cs:565]: Remove newline

File: src/System.CommandLine.Tests/ParserTests.cs [hunk 3]
```diff
@@ -772,42 +775,39 @@ public void Options_only_apply_to_the_nearest_command()
...
-                });
+                    {
+                        new CliCommand("complete")
+                        {
+                            new CliArgument<string>("arg"),
+                            new CliOption<int>("--position")
+                        }
+                    };
+
+            ParseResult result = CliParser.Parse(command, new[] { "the-command",
```
Comment (KathleenDollard) [inline src/System.CommandLine.Tests/ParserTests.cs:806]: Check whitespace

File: src/System.CommandLine.Tests/ParserTests.cs [hunk 4]
```diff
@@ -962,107 +962,108 @@ public void Command_default_argument_value_does_not_override_parsed_value()
...
-                .Name
-                .Should()
-                .Be("the-directory");
+                  .Name
+                  .Should()
+                  .Be("the-directory");
         }
-*/
+        */
+
```
Comment (KathleenDollard) [inline src/System.CommandLine.Tests/ParserTests.cs:978]: Remove blank line

File: src/System.CommandLine.Tests/TokenizerTests.cs [hunk 5]
```diff
@@ -23,6 +23,7 @@ public void The_tokenizer_can_handle_single_option()
             List<CliToken> tokens = null;
             List<string> errors = null;
             Tokenizer.Tokenize(args, command, new CliConfiguration(command), true, out tokens, out errors);
+            Tokenizer.Tokenize(args, command, new CliConfiguration(command), true, out tokens, out errors);
```
Comment (KathleenDollard) [inline src/System.CommandLine.Tests/TokenizerTests.cs:26]: Is this line really doubled?

File: src/System.CommandLine/ParseResult.cs [hunk 6]
```diff
@@ -44,6 +47,7 @@ public sealed class ParseResult
             Configuration = configuration;
             _rootCommandResult = rootCommandResult;
             CommandResult = commandResult;
+            this.valueResultDictionary = valueResults;
```
Comment (KathleenDollard) [inline src/System.CommandLine/ParseResult.cs:50]: Probably do not need this this

File: src/System.CommandLine/ParseResult.cs [hunk 7]
```diff
@@ -159,6 +165,9 @@ public IReadOnlyList<string> UnmatchedTokens
         public override string ToString() => ParseDiagramAction.Diagram(this).ToString();
         */
 
+        public ValueResult? GetValueResult(CliSymbol symbol)
```
Comment (KathleenDollard) [inline src/System.CommandLine/ParseResult.cs:168]: Consider TryGetValue here

File: src/System.CommandLine/Parsing/ArgumentResult.cs [hunk 8]
```diff
@@ -22,13 +22,32 @@ public sealed class ArgumentResult : SymbolResult
...
+        }
+
         /// <summary>
         /// The argument to which the result applies.
         /// </summary>
         public CliArgument Argument { get; }
 
         internal bool ArgumentLimitReached => Argument.Arity.MaximumNumberOfValues == (_tokens?.Count ?? 0);
 
+
```
Comment (KathleenDollard) [inline src/System.CommandLine/Parsing/ArgumentResult.cs:50]: Remove blank line

File: src/System.CommandLine/Parsing/CommandValueResult.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
```
Comment (KathleenDollard) [inline src/System.CommandLine/Parsing/CommandValueResult.cs:4]: Clean up usings

File: src/System.CommandLine/Parsing/SymbolResultTree.cs [hunk 10]
```diff
@@ -60,35 +67,56 @@ internal IEnumerable<SymbolResult> GetChildren(SymbolResult parent)
             }
         }
 
+        internal Dictionary<CliSymbol, ValueResult> GetValueResultDictionary()
+        {
+            var dict = new Dictionary<CliSymbol, ValueResult>();
+            foreach (KeyValuePair<CliSymbol, SymbolResult> pair in this)
+            {               
+                    var result = pair.Value;
```
Comment (KathleenDollard) [inline src/System.CommandLine/Parsing/SymbolResultTree.cs:75]: Whitespace

File: src/System.CommandLine/Parsing/ValueResultExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Comment (KathleenDollard) [inline src/System.CommandLine/Parsing/ValueResultExtensions.cs:1]: Consider adding this code to ArgumentResult

Also, be sure the ValueResultOutcome method is used

**Review by mhutch — APPROVED** [review]

**Pull request merged**